Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -38,3 +42,135 @@ theorem fermatLastTheoremThree_case_1 {a b c : ℕ} (hdvd : ¬ 3 ∣ a * b * c)
+      rw [← hF]
+      exact dvd_add (dvd_pow (by exact_mod_cast ha) (by decide))
+        (dvd_pow (by exact_mod_cast hb) (by decide))
+    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
+  refine Finset.dvd_gcd (fun x hx ↦ ?_)
+  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
+  rcases hx with (hx | hx | hx)
+  · exact hx ▸ ha
+  · exact hx ▸ hb
+  · exact hx ▸ hc
+
+lemma three_dvd_gcd_of_dvd_b_of_dvd_c : 3 ∣ ({a, b, c} : Finset ℕ).gcd id := by
+  have ha : 3 ∣ a := by
+    have : 3 ∣ (a : ℤ) ^ 3 := by
+      replace hF : (a : ℤ) ^ 3 + (b : ℤ) ^ 3 = (c : ℤ) ^ 3 := by exact_mod_cast hF
+      rw [← eq_sub_iff_add_eq] at hF
+      rw [hF]
+      exact dvd_sub (dvd_pow (by exact_mod_cast hc) (by decide))
+            (dvd_pow (by exact_mod_cast hb) (by decide))
+    exact Int.coe_nat_dvd.1 <| Int.prime_three.dvd_of_dvd_pow this
+  refine Finset.dvd_gcd (fun x hx ↦ ?_)
+  simp only [Finset.mem_insert, Finset.mem_singleton] at hx
+  rcases hx with (hx | hx | hx)
+  · exact hx ▸ ha
+  · exact hx ▸ hb
+  · exact hx ▸ hc
+
+open Finset Int Nat in
+/-- To prove `FermatLastTheoremFor 3`, we may assume that `¬ 3 ∣ a`, `¬ 3 ∣ b`, `a` and `b`
+are coprime and `3 ∣ c`. -/
+theorem fermatLastTheoremThree_of_three_dvd_only_c
+    (H : ∀ a b c : ℤ, c ≠ 0 → ¬ 3 ∣ a → ¬ 3 ∣ b  → 3 ∣ c → IsCoprime a b → a ^ 3 + b ^ 3 ≠ c ^ 3) :
+    FermatLastTheoremFor 3 := by
+  refine fermatLastTheoremWith_of_fermatLastTheoremWith_coprime (fun a b c ha hb hc Hgcd hF ↦?_)
+  by_cases h1 : 3 ∣ a * b * c
+  swap; exact fermatLastTheoremThree_case_1 h1 hF
+  rw [(Nat.prime_three).dvd_mul, (Nat.prime_three).dvd_mul] at h1
+  have h3 : ¬(3 ∣ 1) := by decide
+  rcases h1 with ((⟨k, hk⟩ | ⟨k, hk⟩) | ⟨k, hk⟩)
+  · refine H (-(c : ℤ)) b (-(a : ℤ)) (by simp [ha]) (fun hdvd ↦ h3 ?_) (fun hdvd ↦ h3 ?_) ?_ ?_ ?_
+    · exact Hgcd.symm ▸ three_dvd_gcd_of_dvd_a_of_dvd_c ⟨k, hk⟩ (coe_nat_dvd.1 (dvd_neg.1 hdvd)) hF
+    · exact Hgcd.symm ▸ three_dvd_gcd_of_dvd_a_of_dvd_b ⟨k, hk⟩ (by exact_mod_cast hdvd) hF
+    · exact ⟨-k, by simp [hk]⟩
+    · refine (isCoprime_iff_coprime.2 (coprime_of_dvd' (fun p hp hpc hpb ↦ ?_))).neg_left
+      rw [← Hgcd]; refine dvd_gcd (fun x hx ↦ ?_)
+      simp only [mem_insert, mem_singleton] at hx
+      rcases hx with (hx | hx | hx)
+      · refine hx ▸ (hp.dvd_of_dvd_pow <| (Nat.dvd_add_iff_right (m := b ^ 3) (n := a ^ 3)
+          (dvd_pow hpb (by decide))).2 ?_)
+        rw [add_comm, hF]
+        exact dvd_pow hpc (by decide)
+      · exact hx ▸ hpb
+      · exact hx ▸ hpc
+    · rw [Odd.neg_pow (by decide), Odd.neg_pow (by decide), add_comm, ← sub_eq_add_neg,
+        sub_eq_iff_eq_add, add_comm, ← sub_eq_add_neg, eq_sub_iff_add_eq, add_comm]
+      exact_mod_cast hF
+  · refine H a (-(c : ℤ)) ((-(b : ℤ))) (by simp [hb]) (fun hdvd ↦ h3 ?_) (fun hdvd ↦ h3 ?_) ?_ ?_ ?_
+    · exact Hgcd.symm ▸ three_dvd_gcd_of_dvd_a_of_dvd_b (by exact_mod_cast hdvd) ⟨k, hk⟩ hF
```

Comment:
Again, the triple repetition of the proof certainly can be mostly avoided.

Reply:
I've more or less avoided repetition, as you suggested. It's possible one can do better, but this seems reasonable to me.